File: src/sdesheet/Day07/ThreeSum.java
/*
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]]
 * such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
 * 
 * Notice that the solution set must not contain duplicate triplets.
 */

package sdesheet.Day07;

import java.util.*;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        // Sort the array
        Arrays.sort(nums);
        
        // Initialize the array for resultant triplets
        List<List<Integer>> res = new ArrayList<>();
        int len = nums.length;

        // consider a as each eleemnts till the 3rd last element
        for (int i = 0; i < len - 2; i++) {

            // if we are considering 1st element or we are considering an element we have not considered before
            // i.e element at i is not equal elemnt at i - 1
            if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
                int low = i + 1; // low at i + 1
                int high = len - 1; // high at end
                int sum = (0 - nums[i]); // b + c = 0 - a
                while (low < high) {

                    // if we find the right b-c pair, we add it to list
                    if (nums[low] + nums[high] == sum) {
                        ArrayList<Integer> temp = new ArrayList<>();
                        temp.add(nums[i]);
                        temp.add(nums[low]);
                        temp.add(nums[high]);
                        res.add(temp);
                        
                        // We ignore all next duplicate elements
                        while (low < high && nums[low] == nums[low + 1])
                            low++;
                        // We ignore all duplicate elements left to elemnt at high
                        while (low < high && nums[high] == nums[high - 1])
                            high--;
                        low++;
                        high--;
                    }
                    
                    // If the sum of b and c is lower than what we require, we increment low pointer
                    else if (nums[low] + nums[high] < sum)
                        low++;
                    // else we decrement high pointer
                    else 
                        high--;
                }
            }
        }
        
        return res;
    }
}

File: src/neetcode/TwoPointers/three_sum.py
"""
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.
"""


from typing import List


class ThreeSum:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        size = len(nums)
        for i in range(size - 2):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            first = nums[i]
            l, r = i + 1, size - 1
            while l < r:
                curr = nums[l] + nums[r] + first
                if curr < 0:
                    l += 1
                elif curr > 0:
                    r -= 1
                else:
                    res.append([first, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
        return res

"""
for [-3, 0, 2, -3, 2, 1]... we sort it -> [-3, -3, 0, 1, 2, 2]
we take each elemnt from first till 3rd last, and combine with other 2
"""